File: 383-ransom-note/383-ransom-note.scala
import scala.collection.mutable._

object Solution {
    def canConstruct(ransomNote: String, magazine: String): Boolean = {
        var map: Map[Char, Int] = Map().withDefaultValue(0)
        
        magazine.foreach { c => map += {c -> (map(c) + 1)}}
        
        var res = true
        
        ransomNote.foreach { c =>
            if (map(c) > 0) {
                map(c) = map(c) - 1
            } else {
                res = false
            }
        }
        
        res
    }
}

File: 383-ransom-note/383-ransom-note.java
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        Map<Character, Integer> wordCnt = new HashMap<>();
        
        for(char c : magazine.toCharArray()) {
            wordCnt.put(c, wordCnt.getOrDefault(c, 0) + 1);
        }
        
        for(char c : ransomNote.toCharArray()) {
            int cnt = wordCnt.getOrDefault(c, 0);
            if (cnt <= 0) return false;
            wordCnt.put(c, cnt - 1);
        }
        
        return true;
    }
}